src/common/selftest/selftest_firstlayer.cpp

```diff
@@ -0,0 +1,342 @@
+    return LoopResult::RunNext;
+}
+
+/*****************************************************************************/
+// Load
+LoopResult CSelftestPart_FirstLayer::stateFilamentLoadEnqueueGcode() {
+    if (state_selected_by_user != StateSelectedByUser::Load) {
+        return LoopResult::RunNext;
+    }
+
+    queue.enqueue_one_now("M701 W0"); // load, no return no cooldown
+    LogInfo("%s load enqueued", rConfig.partname);
+    return LoopResult::RunNext;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateFilamentLoadWaitFinished() {
+    //we didn't wanted to load, so we are not waiting for anything
+    if (state_selected_by_user != StateSelectedByUser::Load) {
+        return LoopResult::RunNext;
+    }
+    //wait for operation to finish
+    if (filament_gcodes::InProgress::Active()) {
+        LogInfoTimed(log, "%s waiting for load to finish", rConfig.partname);
+        return LoopResult::RunCurrent;
+    }
+    //check if we returned from preheat or finished the load
+    PreheatStatus::Result res = PreheatStatus::ConsumeResult();
+    if (res == PreheatStatus::Result::DoneNoFilament) {
+        // in case it flickers, we might need to add change of state
+        // IPartHandler::SetFsmPhase(PhasesSelftest::);
+        return LoopResult::RunNext;
+    } else {
+        return LoopResult::GoToMark;
+    }
+}
+
+/*****************************************************************************/
+// Unload
+LoopResult CSelftestPart_FirstLayer::stateFilamentUnloadEnqueueGcode() {
+    if (state_selected_by_user != StateSelectedByUser::Unload) {
+        return LoopResult::RunNext;
+    }
+
+    queue.enqueue_one_now("M702 W0"); // unload, no return no cooldown
+    LogInfo("%s unload enqueued", rConfig.partname);
+    return LoopResult::RunNext;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateFilamentUnloadWaitFinished() {
+    //we didn't wanted to unload, so we are not waiting for anything
+    if (state_selected_by_user != StateSelectedByUser::Unload) {
+        return LoopResult::RunNext;
+    }
+    //wait for operation to finish
+    if (filament_gcodes::InProgress::Active()) {
+        LogInfoTimed(log, "%s waiting for unload to finish", rConfig.partname);
+        return LoopResult::RunCurrent;
+    }
+
+    // in case it flickers, we might need to add change of state
+    // IPartHandler::SetFsmPhase(PhasesSelftest::);
+
+    // it does not matter if unload went well or was aborted
+    // we need to ask user what to do in both cases
+    return LoopResult::GoToMark;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateShowCalibrateMsg() {
+    IPartHandler::SetFsmPhase(PhasesSelftest::FirstLayer_calib);
+    return LoopResult::RunNext;
+}
+
+static constexpr int axis_steps_per_unit[] = DEFAULT_AXIS_STEPS_PER_UNIT;
+static constexpr float z_offset_step = 1.0F / float(axis_steps_per_unit[2]);
+static constexpr float nozzle_to_probe[3] = NOZZLE_TO_PROBE_OFFSET;
+static constexpr float z_offset_def = nozzle_to_probe[2];
```

Comment:
I expect there exist macro in Marlin which names Z axis index.

Reply:
there is AxisEnum, will fix it